# 聚合查询


从本篇文章开始我们将步入复杂查询的领域.

数据库管理软件除了承担存储数据的功能(数据的简单增删改查),一般也需要承担一定的数据分析功能,就像一个仓库管理员不光会要管入库出库,也要能清点货物.这种清点货物的能力就是复杂查询的范围.

本文将介绍的聚合查询主要作用是计算一些数据的统计值,比如清点货物的总数,货物的总价值,不同类型货物的总数和总价值这些.

因此大致可以分为两种操作:


1. 对表进行聚合分析

2. 对表进行分组后按分组进行分析

做这类统计计算,这就需要有对应的聚合函数.除此之外也需要对应的SQL语法支持

本篇文章将介绍两种新的语句:

+ `GROUP BY`子句 用于按列中的值分组

+ `HAVING`子句 用于根据条件筛选组


使用他们就可以构造聚合查询了


In [1]:
-- connection: postgres://postgres:postgres@localhost:5432/postgres

In [2]:
-- autocommit: true

switched autocommit mode to True

## 准备工作

本文依然使用商品表来作为例子

> 商品表

商品名|品牌|颜色款式|商品种类|商品价格|进货价格|登记日期
---|---|---|---|---|---|---
ATS001|A|黑白宽条纹款|T恤|59|30|2019-06-02 10:00:00
ATS002|A|蓝白宽条纹款|T恤|79|40|2020-03-02 10:00:00
ATSY001|A|蓝白宽条纹初音限定款|T恤|299|45|2020-04-02 10:00:00
BTS001|B|黑白宽条纹款|T恤|59|30|2020-02-02 10:00:00
BTS002|B|蓝白宽条纹款|T恤|79|40|2020-03-02 10:00:00
BTSY001|B|纯白款|T恤|49|20|2020-04-02 10:00:00
ASS000|A|黑色款|短裤|99|60|2020-03-05 10:00:00
ASS001|A|米色款|短裤|99|60|2020-03-05 10:00:00
ASSY001|A|米色底红色花纹超人限定款|短裤|599|140|2020-04-02 10:00:00
BSS001|B|黑色款|短裤|89|50|2020-03-05 10:00:00
BSS002|B|黑底白花款|短裤|96|60|2019-10-05 10:00:00
BSS003|B|白底红花款|短裤|96|60|2020-03-05 10:00:00

In [3]:
CREATE TABLE IF NOT EXISTS Commodity
(id     serial4    PRIMARY KEY,-- 商品id,
 name    VARCHAR(100) NOT NULL UNIQUE,-- 商品名
 brand   VARCHAR(32)  , -- 品牌
 style   VARCHAR(100) NOT NULL, -- 款式
 type    VARCHAR(32)  NOT NULL, -- 商品类型
 sale_price      INT4      DEFAULT 0, -- 售价
 purchase_price  INT4      , -- 进价
 ctime     TIMESTAMPTZ  DEFAULT (now()) --商品录入的时间,默认为当前时间
)

In [4]:
INSERT INTO Commodity ( 
    name,
    brand,
    style,
    type,
    sale_price,
    purchase_price ,
    ctime
) VALUES (
    'ATS001',
    'A',
    '黑白宽条纹款',
    'T恤',
    59,
    30,
    '2019-06-02T10:00:00.000Z'
),
(
    'ATS002',
    'A',
    '蓝白宽条纹款',
    'T恤',
    79,
    40,
    '2020-03-02T10:00:00.000Z'
),
(
    'ATSY001',
    'A',
    '蓝白宽条纹初音限定款',
    'T恤',
    299,
    45,
    '2020-04-02T10:00:00.000Z'
),
(
    'BTS001',
    'B',
    '黑白宽条纹款',
    'T恤',
    59,
    30,
    '2020-02-02T10:00:00.000Z'
),
(
    'BTS002',
    'B',
    '蓝白宽条纹款',
    'T恤',
    79,
    40,
    '2020-03-02T10:00:00.000Z'
),

(
    'BTSY001',
    'B',
    '纯白款',
    'T恤',
    49,
    20,
    '2020-04-02T10:00:00.000Z'
),
(
    'ASS000',
    'A',
    '黑色款',
    '短裤',
    99,
    60,
    '2020-03-05T10:00:00.000Z'
),
(
    'ASS001',
    'A',
    '米色款',
    '短裤',
    99,
    60,
    '2020-03-05T10:00:00.000Z'
),
(
    'ASSY001',
    'A',
    '米色底红色花纹超人限定款',
    '短裤',
    599,
    140,
    '2020-04-02T10:00:00.000Z'
),


(
    'BSS001',
    'B',
    '黑色款',
    '短裤',
    89,
    50,
    '2020-03-05T10:00:00.000Z'
),
(
    'BSS002',
    'B',
    '黑底白花款',
    '短裤',
    96,
    60,
    '2019-10-05T10:00:00.000Z'
),
(
    'BSS003',
    NULL,
    '白底红花款',
    '短裤',
    96,
    60,
    '2020-03-05T10:00:00.000Z'
)

In [5]:
SELECT * FROM Commodity

12 row(s) returned.


id  name     brand    style                     type      sale_price    purchase_price  ctime
----  -------  -------  ------------------------  ------  ------------  ----------------  -------------------------
   1  ATS001   A        黑白宽条纹款              T恤               59                30  2019-06-02 10:00:00+00:00
   2  ATS002   A        蓝白宽条纹款              T恤               79                40  2020-03-02 10:00:00+00:00
   3  ATSY001  A        蓝白宽条纹初音限定款      T恤              299                45  2020-04-02 10:00:00+00:00
   4  BTS001   B        黑白宽条纹款              T恤               59                30  2020-02-02 10:00:00+00:00
   5  BTS002   B        蓝白宽条纹款              T恤               79                40  2020-03-02 10:00:00+00:00
   6  BTSY001  B        纯白款                    T恤               49                20  2020-04-02 10:00:00+00:00
   7  ASS000   A        黑色款                    短裤              99                60  2020-03-05 10:00:00+00:00
   8  ASS001   A        米色款                    短裤              99                60  2020-03-05 10:00:00+00:00
   9  ASSY001  A        米色底红色花纹超人限定款  短裤             599               140  2020-04-02 10:00:00+00:00
  10  BSS001   B        黑色款                    短裤              89                50  2020-03-05 10:00:00+00:00
  11  BSS002   B        黑底白花款                短裤              96                60  2019-10-05 10:00:00+00:00
  12  BSS003            白底红花款                短裤              96                60  2020-03-05 10:00:00+00:00

## 聚合函数

在前面的文章中我们已经接触过最简单的聚合函数--`count`.通常聚合函数就是将多行汇总为一行的函数.比较常用的如下:

函数|说明
---|---
COUNT|计算表中的记录数
SUM|计算表中数值列中数据的合计值
AVG|计算表中数值列中数据的平均值
VARIANCE|计算表中数值列中的数据方差
MAX|求出表中任意列中数据的最大值
MIN|求出表中任意列中数据的最小值

需要注意的是这6个聚合函数中只有`count`的参数可以是一列或者多列数据,其他5个的参数都必须是一列数据且数据的类型必须是数值型.
不同的数据库管理系统会有不同的聚合函数实现,但上面5个基本都是有的,有特别需求的时候就需要去看具体的数据库管理系统的实现了.


> 查看最小进价短裤的进价

In [6]:
SELECT MIN(purchase_price) FROM Commodity WHERE type='短裤'

1 row(s) returned.


min
-----
   50

> 查看短裤类商品的进价方差

In [7]:
SELECT VARIANCE(purchase_price) FROM Commodity WHERE type='短裤'

1 row(s) returned.


variance
----------
   1136.67

## 对表进行分组

我们如果要知道不同分类的商品的均价,如果不能分组那么就需要知道类型的范围,并用上面的语句查询范围数量次.明显这既不优雅也不易维护.

我们可以使用`GROUP BY`语句来对分类这一列进行分组

其语法是:

```SQL
SELECT
...
FROM
...
[WHERE
 ...
 ]
GROUP BY <列1>,<列2>...
```

含有`GROUP BY`语句的查询会先执行`WHERE`子句中的筛选操作,然后再做分组.

> 查看各类商品的均价

In [8]:
SELECT type,AVG(purchase_price) 
FROM Commodity 
GROUP BY type

2 row(s) returned.


type        avg
------  -------
T恤     34.1667
短裤    71.6667

在`GROUP BY`子句中指定的列称为聚合键或者分组列,他们**不能是别名*.由于能够决定表的切分方式,所以是非常重要的列,通常会要设置索引以提高查询速度.当然`GROUP BY`子句也可以通过逗号分隔指定多列作为分组列.

通常使用`GROUP BY`的聚合查询可以在`SELECT`子句中指定**聚合键**和**分组的聚合查询结果**,**常数或标量子查询**作为结果.

> 查看各类商品不同品牌的进价均价

In [9]:
SELECT type,
brand,
AVG(purchase_price) 
FROM Commodity 
GROUP BY type,brand

5 row(s) returned.


type    brand        avg
------  -------  -------
短裤    B        55
T恤     A        38.3333
T恤     B        30
短裤    A        86.6667
短裤             60

需要注意聚合键中的`NULL`值同样是一个分组,因此要先做好处理

In [10]:
SELECT type,
brand,
AVG(purchase_price) 
FROM Commodity 
WHERE brand IS NOT NULL
GROUP BY type,brand

4 row(s) returned.


type    brand        avg
------  -------  -------
短裤    B        55
T恤     A        38.3333
T恤     B        30
短裤    A        86.6667

## 为聚合结果指定条件


我们可以使用`WHERE`子句筛选行数据,与之对应的是如何筛选分组数据,SQL中使用`HAVING`子句来实现这个功能.其语法是接在`GROUP BY`语句后面指定筛选条件:

```SQL
SELECT
...
FROM
...
[WHERE
 ...
 ]
GROUP BY <列1>,<列2>...
HAVING <条件1> [AND/OR/NOT <条件2> ....]
```
需要注意的是HAVING的条件必须是聚合函数构造的.其中可以存在的值和聚合查询中`SELECT`子句后面的要求一样,即只能是


+ 聚合键
+ 分组的聚合查询结果
+ 常数或标量子查询

但通常我们不会在`HAVING`子句中做对聚合键的筛选,因为这个在`WHERE`子句中就能干,而且通常`WHERE`子句的性能会更好.

> 查询各类商品不同品牌,且进价均价大于35元的的平均售价

In [12]:
SELECT type,
brand,
AVG(sale_price) as avg_sale_price
FROM Commodity 
WHERE brand IS NOT NULL
GROUP BY type,brand
HAVING AVG(purchase_price) >35

3 row(s) returned.


type    brand      avg_sale_price
------  -------  ----------------
短裤    B                  92.5
T恤     A                 145.667
短裤    A                 265.667

## GROUPING运算符

上面我们一斤可以通过`GROUP BY`语句将表按不同的品牌类型进行分类汇总统计了,但另一个很常见的需求*合计*却无法同时实现.实际上要解决这一需求我们可以使用`GROUPING`运算符实现.

`GROUPING`运算符包含以下3种:

+ `ROLLUP`

+ `CUBE`

+ `GROUPING SETS`

上面这个需求就可以直接使用`ROLLUP`来实现.

需要注意`GROUPING运算符`是一个相对较新的语法,并不是所有的数据库管理系统都支持或者完全支持这类运算符,比如MYSQL

### ROLLUP 

`ROLLUP`是'卷起'的意思,比如卷起百叶窗,窗帘卷等等.其名称也形象地说明了该操作能够得到像从小计到合计这样,从最小的聚合级开始,聚合单位逐渐扩大的结果.

`GROUP BY`子句实现的是聚合,但当指定键以后尤其是指定了多个键以后,聚合也是有顺序的,实际上`GROUP BY`子句的聚合顺序是其后面的列从右往左进行聚合的,而`ROLLUP`运算符则可以在每次聚合后记下这些组,然后再一级一级的向上再聚合,直到将全表数据都聚到一起,就向是圈窗帘一样.



为了形象些,我们假设有语句`group by rollup(A,B,C)`,它会按如下顺序执行:

1. 执行`group by A,B,C`获得分组
2. 执行`group by A,B`获得分组
3. 执行`group by A`获得分组
4. 将全量数据作为一个分组


这些由组聚合而成的组也就是上面的2,3,4步骤得到的组被称作**超级分组**

> 计算所有不同品牌不同类型的商品的均价,并计算不同类型商品的均价和全部商品的均价

In [13]:
SELECT type,
brand,
AVG(purchase_price) 
FROM Commodity 
GROUP BY ROLLUP(type,brand)
ORDER BY type,brand

8 row(s) returned.


type    brand        avg
------  -------  -------
T恤     A        38.3333
T恤     B        30
T恤              34.1667
短裤    A        86.6667
短裤    B        55
短裤             60
短裤             71.6667
                 52.9167

#### 使用`GROUPING`函数确定是否是分组

上面的查询我们可以看到`短裤`的合计无法被识别出是60还是71.6667,这是因为品牌数据中我们存在`NULL`值.最好我们可以有一种方法可以识别出这一行究竟是一般的组还是超级分组.这可以使用`GROUPING`函数来确定,当它返回`1`时说明它是超级分组,是`0`时则标识它是一般分组.

In [19]:
SELECT CASE GROUPING(type) WHEN 1 THEN '类型均值' ELSE type END AS type,
 CASE GROUPING(brand) WHEN 1 THEN '品牌均值' ELSE brand END AS brand,
AVG(purchase_price) 
FROM Commodity 
GROUP BY ROLLUP(type,brand)
ORDER BY type,brand

8 row(s) returned.


type      brand         avg
--------  --------  -------
T恤       A         38.3333
T恤       B         30
T恤       品牌均值  34.1667
短裤      A         86.6667
短裤      B         55
短裤      品牌均值  71.6667
短裤                60
类型均值  品牌均值  52.9167

### CUBE

`ROLLUP`在执行时是有信息缺失的,我们可以得到类型的均值但无法同时得到品牌的均值,这是因为它是在一个`平面`上折叠数据


`CUBE`不像`ROLLUP`那么简单粗暴,而是会将超级分组的各个组合都呈现出来,因此可以认为它是一个在`3维空间`中构造数据的操作.

为了直观些,我们还拿上面的例子看

我们假设有语句`group by cube(A,B,C)`,它会按如下顺序执行:

1. 执行`group by A,B,C`获得分组
2. 执行`group by A,B`,`group by B,C`,`group by A,C`获得分组
3. 执行`group by A`,`group by B`获得分组
4. 将全量数据作为一个分组

In [20]:
SELECT CASE GROUPING(type) WHEN 1 THEN '类型均值' ELSE type END AS type,
 CASE GROUPING(brand) WHEN 1 THEN '品牌均值' ELSE brand END AS brand,
AVG(purchase_price) 
FROM Commodity 
GROUP BY CUBE(type,brand)
ORDER BY type,brand

11 row(s) returned.


type      brand         avg
--------  --------  -------
T恤       A         38.3333
T恤       B         30
T恤       品牌均值  34.1667
短裤      A         86.6667
短裤      B         55
短裤      品牌均值  71.6667
短裤                60
类型均值  A         62.5
类型均值  B         40
类型均值  品牌均值  52.9167
类型均值            60

我们只是改动了`ROLLUP`为`CUBE`就同时得到了品牌均值和类型均值.

### GROUPING SETS

`GROUPING SETS`会只构造超级分组而过滤掉其他所有普通分组.


In [21]:
SELECT CASE GROUPING(type) WHEN 1 THEN '类型均值' ELSE type END AS type,
 CASE GROUPING(brand) WHEN 1 THEN '品牌均值' ELSE brand END AS brand,
AVG(purchase_price) 
FROM Commodity 
GROUP BY GROUPING SETS(type,brand)
ORDER BY type,brand

5 row(s) returned.


type      brand         avg
--------  --------  -------
T恤       品牌均值  34.1667
短裤      品牌均值  71.6667
类型均值  A         62.5
类型均值  B         40
类型均值            60

## 收尾

本篇文章介绍的是复杂查询中的聚合查询,聚合查询常用于日常统计,因此对于数据科学相关工作十分有用.

聚合查询语句中以如下语法书写:

```SQL
SELECT 列1 [AS 别名1],
  列2 [AS 别名2], 
  列3 [AS 别名3], 
  ……
FROM <表名>| (<子查询>) AS 别名
[
  WHERE <条件1> [AND/OR/NOT <条件2> ....]
]
[ 
  GROUP BY <列1>,<列2>...
]
[ 
  HAVING <条件1> [AND/OR/NOT <条件2> ....]
]
[
  ORDER BY (列1, 列2, 列3, ……) [DESC]
]
[
  LIMIT <行数> 
  OFFSET <行数>
]
```

其中关键的注意点如下:

+ 聚合函数的结果是一个数值
+ `SELECT`/`HAVING`子句中可以指定的范围是**聚合键**,**分组的聚合查询结果**,**常数或标量子查询**
+ `GROUPING`运算符不同的构造分组的方式.


In [22]:
DROP TABLE IF EXISTS Commodity